Refactor host_repo -> domain errors

louisruch opened the pull request:
### What does this PR do  

Updates domain errors to not print redundant error code info, i.e. only prints error code info if the wrapped error does not exist or is a different code
Removes new line from domain error stringer
Refactors create host path to domain errors
Refactors errors.New -> errors.E
Adds a New and Wrap helper functions

### Testing  

Updated unit tests and did a number of manual tests with errors injected

### Follow-up work 

Refactor remaining host repo

louisruch commented:
@talanknight @jimlambrt @mgaffney PTAL
I added a couple of error codes which give a bit more clear idea of the type of error that occurred let me know thoughts on naming and alternative ideas on this.  
I moved away from using sentials as much as I could, this also means I needed to update a number of tests to do an `error.Match` instead of an `error.Is`
Finally, I made a helper function to strip the line number off the Op in `error_test.go` as I did not see a need to verify the line number in the test as the code is just using stdlib here to get that line number

louisruch commented on internal/host/static/repository_host.go:
```diff
@@ -23,44 +23,45 @@ import (
 func (r *Repository) CreateHost(ctx context.Context, scopeId string, h *Host, opt ...Option) (*Host, error) {
 	if h == nil {
-		return nil, fmt.Errorf("create: static host: %w", errors.ErrInvalidParameter)
+		return nil, errors.New(errors.InvalidParameter, errors.WithMsg("nil host"))
```
So the open question here is do we want to move away at this point from this generic `InvalidParameter` and return a specific `NilHost` error with this PR.  This would be a step to having a clear error code/ID for each possible error returned.

Simialarly below we would then have `NoCatalogId`, `NoScopeId` etc

mgaffney replied:
I think returning `InvalidParameter` is fine for all of the checks except for the `ErrInvalidAddress` check.

louisruch commented on internal/errors/code.go:
```diff
@@ -22,12 +22,15 @@ const (
 
 	// General function errors are reserved Codes 100-999
 	InvalidParameter Code = 100 // InvalidParameter represents and invalid parameter for an operation.
+	GenerateId       Code = 101
```
New codes to avoid a generic `internal` or similar error and to avoid using `unknown`

louisruch commented on internal/servers/controller/handlers/errors.go:
```diff
@@ -182,7 +183,13 @@ func ErrorHandler(logger hclog.Logger) runtime.ErrorHandlerFunc {
+			if dErr := errors.Convert(inErr, false); dErr != nil {
+				// This is a domain error, get message to return on API
+				// Do not return the full dErr.Error() as it may contain un-scrubbed info
+				msg = dErr.Msg
```
Note to reviewers: are we ok exposing this message to end-users as it is not necessarily as well-formatted as the other messages we return https://github.com/hashicorp/boundary/blob/main/internal/servers/controller/handlers/errors.go#L21

mgaffney commented on internal/host/static/repository_host.go:
```diff
@@ -23,44 +23,45 @@ import (
 	h.Address = strings.TrimSpace(h.Address)
 	if len(h.Address) < MinHostAddressLength || len(h.Address) > MaxHostAddressLength {
-		return nil, fmt.Errorf("create: static host: bad address: %w", ErrInvalidAddress)
+		return nil, errors.New(errors.InvalidParameter, errors.WithMsg("bad address"))
```
I think a specific error code/ID should returned for an invalid address since it could be a user input error or a programming error.

mgaffney commented on internal/host/static/repository_host.go:
```diff
@@ -23,44 +23,45 @@ import (
 	if opts.withPublicId != "" {
 		if !strings.HasPrefix(opts.withPublicId, HostPrefix+"_") {
-			return nil, fmt.Errorf("create: static host: passed-in public ID %q has wrong prefix, should be %q: %w", opts.withPublicId, HostPrefix, errors.ErrInvalidPublicId)
+			return nil, errors.New(errors.InvalidParameter,
```
I think `InvalidPublicId` (or something similar) should be a specific error code/ID

jimlambrt commented on internal/servers/controller/handlers/hosts/host_service.go:
```diff
@@ -178,8 +178,8 @@ func (s Service) createInRepo(ctx context.Context, scopeId, catalogId string, it
 	if err != nil {
-		if errors.IsUniqueError(err) || errors.Is(err, errors.ErrNotUnique) {
-			// Push this error through so the error interceptor can interpret it correctly.
+		if dErr := errors.Convert(err, false); dErr != nil {
```
Most excellent.

jimlambrt reviewed (changes requested):
Great work!  I hope my comments are helpful.

louisruch commented on internal/errors/error.go:
```diff
@@ -51,33 +64,44 @@ func New(c Code, opt ...Option) error {
 // and attempt to add a helpful error msg as well. If that's not possible, it
 // will return nil
-func Convert(e error) *Err {
+func Convert(e error, withCallerOp bool) *Err {
```
Yeah so this has changed a bit as withOp has dropped completely now in favor of the unique errorId, update to this PR will be pushed shortly

louisruch commented on internal/errors/code.go:
```diff
@@ -22,12 +22,15 @@ const (
 	// General function errors are reserved Codes 100-999
 	InvalidParameter Code = 100 // InvalidParameter represents and invalid parameter for an operation.
+	GenerateId       Code = 101
+	LookupHost       Code = 102
```
oops thanks!

louisruch commented on internal/errors/option.go:
```diff
@@ -43,6 +43,6 @@ func WithMsg(msg string) Option {
 // the error.
```
Dropping this now

louisruch commented on internal/host/static/public_ids.go:
```diff
@@ -24,7 +25,7 @@ func newHostCatalogId() (string, error) {
 func newHostId() (string, error) {
```
Yeah have that fixed up locally

louisruch commented:
@talanknight @jimlambrt @malnick PTAL
This now includes the entire host repo path not just `repository_host.go`
This makes use of the new errorId field
Work to still do (might be in a new PR) is to pass domain errors back from the handler so it is not overwritten with by `fmt.Errorf` at the last step

louisruch commented on internal/servers/controller/handlers/errors_test.go:
```diff
@@ -97,7 +96,7 @@ func TestApiErrorHandler(t *testing.T) {
 		{
 			name: "Db invalid public id error",
-			err:  fmt.Errorf("test error: %w", errors.ErrInvalidPublicId),
+			err:  errors.New(errors.MissingPublicId, "testid", errors.WithMsg("test msg")),
```
@mgaffney @talanknight @jimlambrt should this be a 400 or 500 as it is?

louisruch commented on internal/servers/controller/handlers/errors_test.go:
```diff
@@ -153,7 +152,70 @@ func TestApiErrorHandler(t *testing.T) {
+			},
+		},
+		{
+			name: "Db invalid address",
```
@mgaffney @talanknight @jimlambrt should this be a 400 or 500 as it is?

louisruch commented:
I pulled back some of the larger refactor work and just focused on the changes to the domain error and refactored just the `CreateHost` path so we can get consensus.  

Sample before/after error message based on these domain error changes:

Before these changes:
```
error="create: static host: parameter violation: error #100
new host id: parameter violation: error #100
missing prefix: parameter violation: error #100"
```

WIth these changes
```
error="create: static host: new host id: missing prefix: parameter violation: error #100"
```

louisruch commented on internal/errors/code.go:
```diff
@@ -22,6 +22,8 @@ const (
 
 	// General function errors are reserved Codes 100-999
 	InvalidParameter Code = 100 // InvalidParameter represents and invalid parameter for an operation.
+	InvalidAddress   Code = 101
```
Note to reviewers I add these are non-generic InvalidParameters because I assume we want to return 400 in these two cases, please confirm?

talanknight reviewed (commented):
The error message looks a lot cleaner!  I'm excited to see how this ends up.

louisruch commented:
@jimlambrt @mgaffney after chatting with Mike I made a couple of changes PTAL when you get a chance

jimlambrt commented on internal/db/id.go:
```diff
@@ -17,12 +17,13 @@ func NewPublicId(prefix string) (string, error) {
 }
 
 func newId(prefix string) (string, error) {
+	const op = errors.Op("db.newId")
```
I think you can just use `const op = "db.newId"`  The `errors.Op` isn't needed and you won't have to cast it when using it later.

jimlambrt commented on internal/host/static/repository_host.go:
```diff
@@ -22,45 +22,50 @@ import (
 	if opts.withPublicId != "" {
 		if !strings.HasPrefix(opts.withPublicId, HostPrefix+"_") {
-			return nil, fmt.Errorf("create: static host: passed-in public ID %q has wrong prefix, should be %q: %w", opts.withPublicId, HostPrefix, errors.ErrInvalidPublicId)
+			return nil, errors.New(
```
Just an observation, that this change may require a change to the caller, if they used `errors.Is(err, errors.ErrInvalidPublicId)` to determine how to handle the error.   If you wrapped the sentinel `errors.ErrInvalidPublicId` then you'd be backward compatible.... if that's important.

louisruch replied:
I could not find anywhere in the code that the sentinels removed in this PR are checked (other than some tests not included in this path), perhaps I missed something.

jimlambrt commented on internal/servers/controller/handlers/hosts/host_service.go:
```diff
@@ -178,11 +177,7 @@ func (s Service) createInRepo(ctx context.Context, scopeId, catalogId string, it
 	}
 	out, err := repo.CreateHost(ctx, scopeId, h)
 	if err != nil {
```
any chance we want to test that all the errors raised by the repo are infact domain errors?

louisruch replied:
I updated the tests in this path to do an `errors.Match`-  I could add an `assertDomainErr` check before?

jimlambrt replied:
The tests are great, but what will happen if something is missed an a non-domain error is returned from the repo?

louisruch replied:
Sorry this was more my misunderstanding of the request.  Are you suggesting I put a convert before the return, as I had it in a previous version of this refactor, and only return err directly if it’s a domain error and otherwise return as it is today?

jimlambrt reviewed (changes requested):
Just one tiny bit about wrapping an error that might need attention.  Overall, this is soooo much better!  ty!

louisruch commented on internal/errors/error.go:
```diff
@@ -48,15 +48,63 @@ func New(c Code, opt ...Option) error {
+		case error:
+			e.Wrapped = arg
+		default:
+			// ignore it
```
We should probably log this somewhere, I am not sure how our logging story works any pointers would be great @mgaffney

malnick commented:
@louisruch Make sure to also include a changelog update with this PR for the specific library we're improving error handling for - thank you!

jefferai commented on internal/errors/error.go:
```diff
@@ -32,13 +32,13 @@ type Err struct {
 // WithWrap() - allows you to specify
 // an error to wrap
-func New(c Code, opt ...Option) error {
+func E(c Code, opt ...Option) error {
```
I think E shouldn't have a required Code parameter. If you are wanting to wrap an error (but with advanced options) then you may not have a new code to provide. If there isn't another error given to wrap, then this should simply fill in a suitable default.

louisruch replied:
Would you imagine there being a new `errors.WithCode` option so the caller can provide a code when calling `E` or just have it be the wrapped error code if provided otherwise `Unknown`?

jefferai replied:
Yeah basically I think either you'd need WithWrap, or you'd need WithCode. Since we don't want to error from an error function, the code, if not specified, could be a default. But it's not strictly a required parameter given that if you wrap an error with this function it already has a code.

louisruch replied:
👍

jefferai commented on internal/errors/error.go:
```diff
@@ -104,18 +139,25 @@ func (e *Err) Error() string {
+	var skipInfo bool
+	var wrapped *Err
+	if As(e.Wrapped, &wrapped) {
+		// if wrapped error code is the same as this error, don't print redundant info
```
What if it's the same code but different message/kind/op? This makes an assumption that although it's _possible_ to specify the same code when wrapping an error (since we are supporting multiple codes) that only one place in the stack will actually do so with the same error code. I don't think that's a good assumption. If people can specify codes they very well might...

louisruch replied:
`Message` and `Op` are always printed if not empty, `Code` was specifically left out, with this check, to avoid the following when the stringer is invoked:

```
"static.CreateHost: invalid parameter: parameter violation: error #100 static.newHostId: invalid parameter: parameter violation: error #100 db.newId: missing prefix: parameter violation: error #100"
```

Instead, we get:

```
static.CreateHost: static.newHostId: db.newId: missing prefix: parameter violation: error #100
```

The `parameter violation: error #100` part, which comes from https://github.com/hashicorp/boundary/blob/main/internal/errors/info.go#L15 is printed each time we recursively print the wrapped error

Kind is a field based on code within the info map, I am not sure how we can have the same code but a different Kind?

jefferai reviewed (commented):
Made a couple comments, I think it's looking pretty good!

jimlambrt commented on internal/errors/error.go:
```diff
@@ -32,13 +32,13 @@ type Err struct {
 	Wrapped error
```
The formatting of this comment doesn't quite render correctly as a list for the supported options.  Happy to help with suggested fixes, but it's not very readable at the moment when rendered as godoc.

jimlambrt commented on internal/errors/info.go:
```diff
@@ -16,6 +16,22 @@ var errorCodeInfo = map[Code]Info{
 		Message: "invalid parameter",
```
Can you add a godoc comment for the Info struct above?  ty!

jimlambrt reviewed (changes requested):
Looks really good!  Ty for your patience with all of us as we aligned on something we wanted to live with for a long time.  Just a few minor comments...

jefferai commented on internal/errors/is.go:
```diff
@@ -89,3 +89,20 @@ func IsMissingTableError(err error) bool {
+
+// IsNotFoundError returns a boolean indicating whether the error is known to
+// report a not found violation.
+func IsNotFoundError(err error) bool {
```
This function illustrates one of my concerns with the multi-Code model. Switching logic that switches on Code (via using functions like this) can have different behavior depending on the ordering in which the codes are tested in the switch. At the same time, this will return if any error in the chain is RecordNotFound, but not necessarily if the originating error is, or even the most recent error. That means behavior of a switch on Code is also dependent on changes to logic that might seem innocuous -- e.g. if one of the wrapped errors decides to add or modify the code it's wrapping, that can also affect the behavior of this function.

louisruch replied:
The current implementation of this `IsX` functions only check if the most recent domain error's code is of a certain type not any in the chain, @mgaffney mentioned in the call yesterday that perhaps we should have the ability to check if any in the chain match which I did give some thought to but have not implemented yet

jefferai replied:
Or we could just not support multiple Codes at this time and only add it in if/when needed. There was a question a few days ago about whether or not we can think of a time when we'd want to supply more than one Code. Did we ever think of any?

louisruch replied:
So the only time we currently do this is when updating/creating a Host and we get a db `NotNull` or `ConstraintViolation` for both these cases the error returned becomes one with a code of `InvalidAddress`
https://github.com/hashicorp/boundary/blob/e8cf4885e5450b74bbcfd13b4a191dd1d40953e5/internal/host/static/repository_host.go#L178

talanknight replied:
We might also do something similar when authenticating (not found error for auth method or login name could be translated into an unauthenticated error potentially)

jefferai replied:
In that case don't we have a single Code anyways? This is the point where we're translating the DB errors to API errors. If two DB errors map to the same API Code, it's still a single Code right?

louisruch replied:
Well the top level would have one code, and therefore the `IsX` function, as it is currently written, will only be satisfied if the top level is `X`.  But the wrapped errors could have a different code:
`{"Code": InvalidAddress, "Msg": "Top level message", "Wrapped": {"Code": "NotNull", "Msg": "Db level message"}}`

louisruch replied:
Currently, the only time the wrapped errors are used are in the stringer

mgaffney commented on internal/errors/is.go:
```diff
@@ -89,3 +89,20 @@ func IsMissingTableError(err error) bool {
+		return false
+	}
+
+	var domainErr *Err
+	if errors.As(err, &domainErr) {
+		if domainErr.Code == RecordNotFound {
+			return true
+		}
+	}
```
Could this be simplified to:
```go
return errors.Is(err, ErrRecordNotFound)
```

louisruch replied:
No, because not all `RecordNotFound` codes are the sentinel `ErrRecordNotFound`.  As I am doing this refactor I am also removing the dependence on the sentinels, at least I understood that to be one of the objectives...removing them does however cause this, @jimlambrt also previously brought up the benefit of wrapping the sentinel does allow for the `Is` check.  We have this comment in: https://github.com/hashicorp/boundary/blob/main/internal/errors/sentinels.go#L4
```
Creating new Sentinel type errors like these should be deprecated in favor of the new Err type that includes unique Codes and a Matching function.
```
If we are planning to not create new sentinels I see this as also wanting to remove dependence on old sentinels...

talanknight replied:
Keep in mind also that As() will stop at the top level domain error while Is() will check every error down the wrapped list until it finds one that results in Is() being true or it reaches the end.

louisruch commented:
Most recent changes can be viewed in these commits: https://github.com/hashicorp/boundary/pull/786/files/3bdc6e958fc745b65d7cd0ead5150dfac0937fd8..13c2f62154d231d8349cf3feff74e4607e601645

louisruch commented:
@jefferai @jimlambrt @talanknight PTAL I am considering unless someone objects, to merge this PR as is and then do the handler refactor in a quick follow up.  I am trying to avoid having this PR never stop growing.  The current state keeps the existing behavior with the modifications in the error package we have discussed.  For a reference to changes since last reviewed:
https://github.com/hashicorp/boundary/pull/786/files/3bdc6e958fc745b65d7cd0ead5150dfac0937fd8..13c2f62154d231d8349cf3feff74e4607e601645

jimlambrt commented on internal/errors/code.go:
```diff
@@ -21,13 +21,17 @@ const (
 	// General function errors are reserved Codes 100-999
-	InvalidParameter Code = 100 // InvalidParameter represents and invalid parameter for an operation.
+	InvalidParameter Code = 100 // InvalidParameter represents an invalid parameter for an operation.
+	InvalidAddress   Code = 101 // InvalidAddress represents an invalid host address for an operation
```
Nit: there are several undocumented exported errors here.

jimlambrt reviewed (dismissed):
Looks great.  One little nit that you can ignore if you wish.
